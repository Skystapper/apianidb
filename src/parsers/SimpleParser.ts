import MapField2 from './field_types/MapFieldStatic'
export default class SimpleParser<V extends { [x: string]: (...args: Array<any>)=>any } > {
  parseData: (data: string)=> { [Property in keyof V]: ReturnType<V[Property]> }

  constructor (response_definition: V) {
    this.parseData = MapField2(response_definition, '|')
  }

}