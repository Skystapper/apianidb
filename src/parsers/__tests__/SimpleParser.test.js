import SimpleParser from '../SimpleParser'

test('parseData', () => {

  const parser = new SimpleParser({
    'number_field': Number,
    'string_field': String,
    'boolean_field': Boolean,
    'custom_field': value => ({ [value]: 'test' }),
  })

  const result = parser.parseData('123|test|0|my_value')
  expect(result).toMatchSnapshot()
})

// test may be redoundant!
test('parseData incoreect field number', () => {
  const parser = new SimpleParser({
    'one': String,
    'two': String,
    'three': String,
  })

  function failToParse () {
    parser.parseData('1|2')
  }
  expect(failToParse).toThrowError(new Error(`Definition and input item length missmatch input: 2 definition: 3`))
})