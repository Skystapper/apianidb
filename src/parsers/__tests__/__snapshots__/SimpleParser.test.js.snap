// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseData 1`] = `
{
  "boolean_field": true,
  "custom_field": {
    "my_value": "test",
  },
  "number_field": 123,
  "string_field": "test",
}
`;
